### PR TITLE
Allow macOS to build LLVM as shared library

### PR DESCRIPTION
**Pull request by topjohnwu**
Inspired by how [homebrew](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/llvm.rb) builds and distributes llvm, here we manually create a symlink with a versioned dylib path to make `llvm-config` work properly. Note, the resulting `rustc` executable and `librustc_driver-<hash>.dylib` still links to the un-versioned `libLLVM.dylib` as expected when distributed in the final output. I have confirmed this by checking `otool -L` on both binaries.

After the change, enabling `llvm.link-shared` and `llvm.thin-lto` will be possible on macOS.

### CODE REVIEW

File: src/bootstrap/native.rs [hunk 1]
```diff
@@ -239,7 +239,7 @@ impl Step for Llvm {
 
         builder.update_submodule(&Path::new("src").join("llvm-project"));
         if builder.llvm_link_shared()
-            && (target.contains("windows") || target.contains("apple-darwin"))
+            && (target.contains("windows") || (target.contains("apple-darwin") && !cfg!(unix)))
```
Comment (jyn514) [inline src/bootstrap/native.rs:242]: `target.contains("apple-darwin") && !cfg!(unix)))`

When will both of these things be true? Shouldn't we just remove apple-darwin?

File: src/bootstrap/native.rs [hunk 2]
```diff
@@ -465,6 +470,30 @@ impl Step for Llvm {
 
         cfg.build();
 
+        // When building LLVM with LLVM_LINK_LLVM_DYLIB for macOS, an unversioned
+        // libLLVM.dylib will be built. However, llvm-config will still look
+        // for a versioned path like libLLVM-14.dylib. Manually create a symbolic
+        // link to make llvm-config happy.
+        #[cfg(unix)]
```
Comment (jyn514) [inline src/bootstrap/native.rs:477]: Please don't add cfgs like this - looking at the target specifications is enough. We want this to work when cross compiling.

**Review by jyn514 — CHANGES_REQUESTED** [review]
How likely is this to break? Has homebrew ever had to change the way they do this?

File: src/bootstrap/native.rs (hunk 2 above)
Comment (topjohnwu) [reply src/bootstrap/native.rs:477]: The reason here is that it relies on `std::os::unix::symlink`, which is only available on Unix platforms

**Comment by Mark-Simulacrum** [comment]
Has this been reported upstream to LLVM? Is there a reason to fix this in our code, vs. in LLVM?

File: src/bootstrap/native.rs (hunk 2 above)
Comment (jyn514) [reply src/bootstrap/native.rs:477]: The windows equivalent is https://doc.rust-lang.org/stable/std/os/windows/fs/fn.symlink_file.html.

**Comment by topjohnwu** [comment]
> Has this been reported upstream to LLVM? Is there a reason to fix this in our code, vs. in LLVM?

Yup, you can check out this [comment](https://github.com/Homebrew/homebrew-core/pull/47146#issuecomment-558106394). There is no response in upstream LLVM, which is why Homebrew decided to workaround the issue instead.

File: src/bootstrap/native.rs (hunk 2 above)
Comment (topjohnwu) [reply src/bootstrap/native.rs]: I have added a branch for cross compiling on Windows and will panic if the build script is not running on either Windows or Unix.

File: src/bootstrap/native.rs [hunk 3]
```diff
@@ -465,6 +468,38 @@ impl Step for Llvm {
...
+
+            // The reason why we build the library path from llvm-config is because
+            // the output of llvm-config depends on its location in the file system.
+            // Make sure we create the symlink exactly where it's needed.
+            let llvm_base = build_llvm_config.parent().unwrap().parent().unwrap();
+            let lib_llvm = llvm_base.join(lib_name);
+            #[cfg(windows)]
+            if std::os::windows::fs::symlink_file("libLLVM.dylib", &lib_llvm).is_err() {
+                panic!("unable to create symbolic link {:?}", lib_llvm.file_name());
+            }
```
Comment (jyn514) [inline src/bootstrap/native.rs:492]:
```suggestion
            t!(std::os::windows::fs::symlink_file("libLLVM.dylib", &lib_llvm), format!("unable to create symbolic link {:?}", lib_llvm.file_name()));
```

File: src/bootstrap/native.rs [hunk 4]
```diff
@@ -465,6 +468,38 @@ impl Step for Llvm {
...
+            #[cfg(unix)]
+            if std::os::unix::fs::symlink("libLLVM.dylib", &lib_llvm).is_err() {
+                panic!("unable to create symbolic link {:?}", lib_llvm.file_name());
+            }
+            #[cfg(all(not(windows), not(unix)))]
+            {
+                panic!("unable to cross-compile target {}", target.triple);
+            }
```
Comment (jyn514) [inline src/bootstrap/native.rs:500]: nit: this is the second place we symlink files, maybe add a `builder.symlink` helper? but this is fine too if it's a hassle

https://github.com/rust-lang/rust/blob/43347397f7c5ca9a670a3bb3890c7187e24a52ab/src/bootstrap/lib.rs#L1449

**Review by jyn514 — APPROVED** [review]
LGTM except for the nits

File: src/bootstrap/native.rs (hunk 4 above)
Comment (topjohnwu) [reply src/bootstrap/native.rs]: Created a helper function as suggested

**Review by jyn514 — CHANGES_REQUESTED** [review]
r=me with the dry_run check added

File: src/bootstrap/native.rs [hunk 5]
```diff
@@ -425,18 +425,21 @@ impl Step for Llvm {
             );
         }
 
+        let mut llvm_suffix: Option<String> = None;
         if let Some(ref suffix) = builder.config.llvm_version_suffix {
             // Allow version-suffix="" to not define a version suffix at all.
             if !suffix.is_empty() {
-                cfg.define("LLVM_VERSION_SUFFIX", suffix);
+                llvm_suffix = Some(suffix.to_string());
```
[the remark this answers is not in the document]
Comment (topjohnwu) [reply src/bootstrap/native.rs]: Adjusted accordingly

**Pull request merged**